feat: Flex - Removed required field false positive for stop_id + added Foreign key violation for location_groups_id

jcpitre opened the pull request:
Closes #1749 

**Summary:**

> foreign_key_violation is triggered when location_group_id is defined in stop_times.txt but not in location_groups.txt. 

- Added a foreign key annotation to location_group_id in stopTimeSchema.

> Check that missing_required_field no longer triggers when there's no stop_id in stop_times.txt AND a present id in locations.geojson OR location_group_id in location_groups.txt. 

- Made the stop_id conditionally required. 
- Added missing fields in stoptimeSchema. 
- Added a validator that checks that no more than one of stop_id, location_goup_id and or stop_id is present
- the presence in locations.geojson is out of scope for this PR.

Also verified that feature `Zone-based Demand Responsive Transit` is present in the list of features for a dataset with Flex. We could not test this in #1776 until the @required was removed from stopId.

**Expected behavior:** 

Most of missing required fields related to stopId should disappear from reports.
Some other errors or warning that were masked by the missing required field notice should appear. Some of these will have to be corrected other issues. e.g. #1755 

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [ ] Add or update any needed [documentation](https://github.com/MobilityData/gtfs-validator/tree/master/docs) to the repo 
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)

davidgamez commented on core/src/main/java/org/mobilitydata/gtfsvalidator/notice/ForbiddenGeographyIdNotice.java:
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2020 Google LLC
```
[picky]: 
```suggestion
 * Copyright 2024 MobilityData
```

davidgamez commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsStopTimeSchema.java:
```diff
@@ -67,6 +73,12 @@ public interface GtfsStopTimeSchema extends GtfsEntity {
   @CachedField
   String stopHeadsign();
 
+  @ConditionallyRequired
+  GtfsTime startPickupDropOffWindow();
+
+  @ConditionallyRequired
+  GtfsTime endPickupDropOffWindow();
```
Is there any validator implementing the conditional required logic for these fields? If not, please remove the annotation and bring it back when the logic is implemented.

jcpitre commented:
About the faliing check

- New errors:
   -  missing_trip_edge: Related to #1755
- Dropped errors:
   - missing_required_field: corrected by this PR
   -  trip_distance_exceeds_shape_distance: Not sure why it dropped TBD
  
- New warnings:
   -    TBD. One problem with the stop_id required field false positive fixed is that stop_times.txt is not considered unparseable any more. This could make some of the warnings appear because stop_times.txt is now parseable.

emmambd commented:
@jcpitre These changes look to be expected and positive! Learnings from this: 

- I've added #1755 back into the 6.0 release since this is a false positive we'll need to correct beforehand. Outside the scope of this PR. 

- In general, I'm wondering about `gb-unknown-uk-aggregate-feed-gtfs-2014` in the performance tests. I'm assuming its performance issues are unrelated to the coding changes, but taking 7 extra minutes to run the validation report seems surprising. 

     Comparing the two most recent validation reports on prod of the Mobility Database: https://mobilitydatabase.org/feeds/mdb-2014 it looks like there's 34,000 fewer trips in the last dataset...not sure if we're using prod data in the acceptance tests. 

- Before these acceptance tests were run, I noticed only 1 feed in our analytics dashboard that had the `missing_required_field` notice because of stop_id and was unrelated to Flex data: https://mobilitydatabase.org/feeds/mdb-1000. It looks like it's a partial GTFS file (doesn't include all required files). This is a feed I would've expected to be affected by the new notice since `stop_id` is no longer included in `missing_required_field`. What am I missing?

emmambd commented:
@jcpitre I also just want to clarify the merging order of these PRs. 

It looks like this can be merged independently on #1805 and #1810? Should adding the checks in the new validator that evaluates the unique id for geography in`locations.geojson` be a new issue or included in #1810?

jcpitre commented:
> @jcpitre I also just want to clarify the merging order of these PRs.
> 
> It looks like this can be merged independently on #1805 and #1810? Should adding the checks in the new validator that evaluates the unique id for geography in`locations.geojson` be a new issue or included in #1810?
Do you mean this: https://github.com/MobilityData/gtfs-validator/blob/6f82f0368c95322d89939a62b680d4d207b9552d/main/src/main/java/org/mobilitydata/gtfsvalidator/validator/GtfsGeojsonFeatureUniqueLocationIdValidator.java
I don't see why it should have its own issue. It's kind of done in #1810 as a demonstration. Also now that location_groups.txt has been added I can just modify the validator to take it into account.

emmambd commented:
Cool! @jcpitre so then it could be added as part of this PR?

jcpitre commented:
> * In general, I'm wondering about `gb-unknown-uk-aggregate-feed-gtfs-2014` in the performance tests. I'm assuming its performance issues are unrelated to the coding changes, but taking 7 extra minutes to run the validation report seems surprising.
The stop_times in this datataset has nearly 42 million records. I added 5 new fields in the schema for that file. So at least for memory it's quite a chunk more, and it takes more to load. The new validator has to go through these 42 million records. Does it amount to 7 minutes? It could I guess.

jcpitre commented:
With the new 16G heap the acceptance tests seem to run properly.
